### PR TITLE
chips: apollo3: Handle the errata for clock double increments

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

[Errata 4.22](https://ambiq.com/wp-content/uploads/2022/01/Apollo3-Blue-Errata-List.pdf): Sometimes the clock can increment twice. This means the timer occurs earlier then actually requested.

From testing this scaling results in the correct time, so we scale the requested ticks to give us an accurate alarm.

### Testing Strategy

Using Tock as a LoRaWAN device

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Comment by lschuermann** [comment]
I'm fine with this change if it works, but I don't quite comprehend what's going on here. In this fix, the `dt` value is recaled to 96.8% of what it was before, right? Is there any special meaning behind this number?

I'm curious why the recommended workaround in the silicon errata doesn't work:

> A second compare whose value is one count higher than the initial compare value should be established. This second compare will ensure that if the first compare is missed because of a glitch-

File: chips/apollo3/src/stimer.rs [hunk 1]
```diff
@@ -170,7 +170,12 @@ impl<'a> Alarm<'a> for STimer<'a> {
     fn set_alarm(&self, reference: Self::Ticks, dt: Self::Ticks) {
         let regs = self.registers;
         let now = self.now();
-        let mut expire = reference.wrapping_add(dt);
+        // Errata 4.22: Sometimes the clock can increment twice
+        // This means the timer occurs earlier then actually requested
+        // From testing this scaling results in the correct time, so we
+        // scale the requested ticks to give us an accurate alarm.
+        let scaled_time = Self::Ticks::from((dt.into_u32() * 1000) / (1000 - 32));
```
Comment (lschuermann) [inline chips/apollo3/src/stimer.rs:177]: This can overflow. We should probably cast to an intermediate u64 to do this re-scaling.

**Review by lschuermann — CHANGES_REQUESTED** [review]

**Comment by alistair23** [comment]
> I'm fine with this change if it works, but I don't quite comprehend what's going on here. In this fix, the `dt` value is recaled to 96.8% of what it was before, right? Is there any special meaning behind this number?
> 
> I'm curious why the recommended workaround in the silicon errata doesn't work:

We already have the workaround from the errata in the Tock code. That ensures that we don't miss an interrupt.

The issue is that the interrupt actually occurs early as the clock is double counting. So we end up waking up too early. This change just scales the delay to be a little longer, ensuring we wake up at the correct time.

**Comment by lschuermann** [comment]
> The issue is that the interrupt actually occurs **early** as the clock is double counting. So we end up waking up too early. This change just scales the delay to be a little longer, ensuring we wake up at the correct time.

Huh, okay, its somewhat surprising that the issue is so prevalent that it would account for a 3.2% reduction in effective tick rate. That's a pretty serious silicon bug to slip through. :)

**Review by lschuermann — APPROVED** [review]

**Comment by alistair23** [comment]
> Huh, okay, its somewhat surprising that the issue is so prevalent that it would account for a 3.2% reduction in effective tick rate. That's a pretty serious silicon bug to slip through. 🫢

The official HAL doesn't use this for delays (they have some ROM function they call) so most users would never notice it

**Comment by lschuermann** [comment]
> The official HAL doesn't use this for delays (they have some ROM function they call) so most users would never notice it

And I presume we can't get access to this ROM to know how they're working around this issue, as it's closed?

The way I understand this fix is that now the alarm is approximately correct, but I doubt that the underlying hardware timer ticks _exactly_ 32/1000 faster than its reference frequency, right? So it'll diverge at some point?

Pinging @ppannuto for final approval & merge, as I know he also uses this HW platform.

**Comment by alistair23** [comment]
> And I presume we can't get access to this ROM to know how they're working around this issue, as it's closed?

Exactly

> The way I understand this fix is that now the alarm is approximately correct, but I doubt that the underlying hardware timer ticks _exactly_ 32/1000 faster than its reference frequency, right? So it'll diverge at some point?

I will still diverge, but a lot slower then it currently does

**Pull request merged**